### PR TITLE
[ntuple] (WIP) RNTuple page cache

### PR DESCRIPTION
**Pull request by mxxo**
Profiles of RNTuple benchmarks (`iotools/cms, lhcb`) showed ~10-20% of
total runtime is due to allocations in `RPageSource::UnsealPage`.

One typical RNTuple use case is an event loop over a fixed number of
columns. Columns are divided into pages with a fixed size, e.g.

```
Events | Column 0 | Column 1 | Column 2
-------+----------+----------+----------
  1-10   p0(40kB)   p1(80kB)   p2(80kB)
 11-20   p3(40kB)   p4(80kB)   p5(80kB)
  ...      ...        ...        ...
```

Before this change, page allocations were performed every time a page
is flushed from a column:

```
Events | Column 0 | Column 1 | Column 2
-------+----------+----------+----------
         [alloc]    [alloc]    [alloc]
  1-10   p0(40kB)   p1(80kB)   p2(80kB)
        [dealloc]  [dealloc]  [dealloc]
         [alloc]    [alloc]    [alloc]
 11-20   p3(40kB)   p4(80kB)   p5(80kB)
  ...      ...        ...        ...
        [dealloc]  [dealloc]  [dealloc]
```

After this change, allocations are held onto and reused if possible
(up to a certain limit, which for now is hard-coded to be 32 pages).

```
Events | Column 0 | Column 1 | Column 2
-------+----------+----------+----------
         [alloc]    [alloc]    [alloc]
  1-10   p0(40kB)   p1(80kB)   p2(80kB)
 11-20   p3(40kB)   p4(80kB)   p5(80kB)
  ...      ...        ...        ...
        [dealloc]  [dealloc]  [dealloc]
```

On SSD, the following serial results were obtained for the lhcb and cms
benchmarks (average of ten cold runs, 95% confidence interval):

```
cms

Before: 105145us +- 1338us
After:   93669us +- 2077us
%diff:  10.9%

lhcb

Before: 2111977us +-  8371us
After:  1770781us +- 11185us
%diff:  16.2%
```

Todo: 
- [x] Handle non-mappable case in `PopulatePage`
- [x] Investigate multi-threaded case

For multithreading, the current implementation just uses the existing `UnsealPage` mechanism and is unaffected by these changes -- deleted pages are returned to the cache which deallocates them properly. 

I tried adding an internal mutex to `RPageCache` and bumping the number of cached pages to 128 on the benchmark machine but the runtime was about the same as the existing cache-less implementation. Further cache size increases did not show improvements, but as expected only having 32 cached pages for 128 threads was much slower.

```
./lhcb -m (64-cores, 10 cold runs, 95% confidence interval)
no cache: 776188us +- 9918us
cache (128):    780214us +- 11888us
%diff: -0.5%
cache (32):    1369689us +- 33365us
%diff: -76.46%
```

### CODE REVIEW

**Comment by jalopezg-r00t** [comment]
@mxxo, as I said in the ROOT I/O meeting -and if you don't mind-, I can take this :-).  Therefore, I have self-assigned this PR.

**Comment by mxxo** [comment]
Hey @jalopezg-r00t! I'm very happy to pass this to your capable hands :) Cheers!

File: tree/ntuple/v7/src/RPageStorageFile.cxx [hunk 1]
```diff
@@ -192,6 +193,85 @@ void ROOT::Experimental::Detail::RPageSinkFile::ReleasePage(RPage &page)
...
+   /// Return a page to the cache for reuse. Page memory ownership is passed to the cache.
+   /// If the cache is full, the smallest memory buffer is deallocated.
+   void ReturnPage(const RPage &page) {
+      fPages.push_back(RecyclePage(page));
+      if (fPages.size() <= 32) {
+         return;
+      }
+      // If the cache is getting too large, deallocate the smallest page
+      auto min_elt = std::min_element(fPages.begin(), fPages.end(),
+         [](const auto &a, const auto &b) { return a.GetCapacity() < b.GetCapacity(); });
```
Comment (jalopezg-r00t) [inline tree/ntuple/v7/src/RPageStorageFile.cxx:233]: I think that sequential search should not hurt much as caches are typically not large, but I will perhaps revisit this also.

File: tree/ntuple/v7/src/RPageStorageFile.cxx [hunk 2]
```diff
@@ -192,6 +193,85 @@ void ROOT::Experimental::Detail::RPageSinkFile::ReleasePage(RPage &page)
...
+   }
+
+   /// Request a page from the cache with at least the specified `capacity`. Ownership of the
+   /// page memory is passed to the caller. If there isn't a suitable page in the cache, a
+   /// new allocation is made. The caller is responsible for properly setting the page's
+   /// column id and element size.
+   RPage RequestPage(std::size_t capacity) {
+      auto it = std::find_if(fPages.begin(), fPages.end(), [&](const auto &p) {
+         return p.GetCapacity() >= capacity;
+      });
```
Comment (jalopezg-r00t) [inline tree/ntuple/v7/src/RPageStorageFile.cxx:245]: As said above, we might also benefit here of using a slightly different approach, e.g. ordered list, or a bucket-based data structure.

File: tree/ntuple/v7/src/RPageStorageFile.cxx [hunk 3]
```diff
@@ -200,11 +280,20 @@ ROOT::Experimental::Detail::RPage ROOT::Experimental::Detail::RPageAllocatorFile
    return newPage;
 }
 
+ROOT::Experimental::Detail::RPage ROOT::Experimental::Detail::RPageAllocatorFile::NewPage(
+   ColumnId_t columnId, std::size_t elementSize, std::size_t nElements)
+{
+   RPage recycledPage = fPageCache->RequestPage(elementSize * nElements);
+   RPage newPage(columnId, recycledPage.GetBuffer(), recycledPage.GetCapacity(), elementSize);
```
Comment (jalopezg-r00t) [inline tree/ntuple/v7/src/RPageStorageFile.cxx:287]: Maybe we could just reuse the returned `RPage` instance from the cache?

File: tree/ntuple/v7/src/RPageStorageFile.cxx [hunk 4]
```diff
@@ -200,11 +280,20 @@ ROOT::Experimental::Detail::RPage ROOT::Experimental::Detail::RPageAllocatorFile
...
+   newPage.TryGrow(nElements);
```
Comment (jalopezg-r00t) [inline tree/ntuple/v7/src/RPageStorageFile.cxx:288]: This should always succeed; add assertion, e.g.
```suggestion
   assert(newPage.TryGrow(nElements) != nullptr);
```

File: tree/ntuple/v7/src/RPageStorageFile.cxx [hunk 5]
```diff
@@ -309,20 +398,35 @@ ROOT::Experimental::Detail::RPage ROOT::Experimental::Detail::RPageSourceFile::P
...
+         // may decide to implement to not use UnsealPage but to custom mapping / decompression code.
+         // Note that usually pages are compressed.
+         memcpy(newPage.GetBuffer(), sealedPageBuffer, bytesPacked);
+      }
+      if (!element->IsMappable()) {
+         auto tmp = fPageAllocator->NewPage(columnId, elementSize, nElts);
+         element->Unpack(tmp.GetBuffer(), newPage.GetBuffer(), nElts);
+         fPageAllocator->DeletePage(newPage);
+         newPage = tmp;
+      }
```
Comment (jalopezg-r00t) [inline tree/ntuple/v7/src/RPageStorageFile.cxx:419]: We should reuse the logic in `UnsealPage()`.  Instead, its implementation should be probable pagecache-aware.
We'll see how to do this after the vacation period.

File: tree/ntuple/v7/src/RPageStorageFile.cxx [hunk 6]
```diff
@@ -309,20 +398,35 @@ ROOT::Experimental::Detail::RPage ROOT::Experimental::Detail::RPageSourceFile::P
...
       fCounters->fSzUnzip.Add(elementSize * pageInfo.fNElements);
    }
 
    const auto indexOffset = clusterDescriptor.GetColumnRange(columnId).fFirstElementIndex;
-   auto newPage = fPageAllocator->NewPage(columnId, pageBuffer.release(), elementSize, pageInfo.fNElements);
    newPage.SetWindow(indexOffset + pageInfo.fFirstInPage, RPage::RClusterInfo(clusterId, indexOffset));
+   // Thread safety (fPageAllocator->DeletePage): fPagePool operations are protected by a mutex.
```
Comment (jalopezg-r00t) [inline tree/ntuple/v7/src/RPageStorageFile.cxx:425]: See the multi-thread comments above.

**Review by jalopezg-r00t — COMMENTED** [review]
Nice work, @mxxo.  I will address all my comments after the vacation period.
Thanks!  Much appreciated!  :-)

**Comment by jalopezg-r00t** [comment]
Hey @mxxo;  I did not forget about this! ;-)  My plans are to retake this in mid-February or so...

**Comment by mxxo** [comment]
Hey @jalopezg-r00t, that's great! It will be cool to see how it turns out :)

For what it's worth I seem to remember not seeing the same improvement on spinning disk while benchmarking. But I have no idea of why that would be the case and it might have just been a setup error by me.

**Comment by jalopezg-r00t** [comment]
> Profiles of RNTuple benchmarks (`iotools/cms, lhcb`) showed ~10-20% of
> total runtime is due to allocations in `RPageSource::UnsealPage`.

@mxxo @jblomer This work will be resumed soon as we suspect that not reusing addresses is affecting RDMA data transfers (which has an impact on the measured throughput in the DAOS backend).  Specifically, we need to confirm that not reusing addresses that appear in the IOVs array in `daos_obj_{fetch,update}()` is related to a higher overhead due to RDMA MR registration.

**Comment by jalopezg-git** [comment]
Superseded by PR #12279 (closing this draft, in principle).  Many thanks for the initial idea. @mxxo!  :tada:

**Comment by jalopezg-git** [comment]
> @mxxo @jblomer This work will be resumed soon as we suspect that not reusing addresses is affecting RDMA data transfers (which has an impact on the measured throughput in the DAOS backend). Specifically, we need to confirm that not reusing addresses that appear in the IOVs array in `daos_obj_{fetch,update}()` is related to a higher overhead due to RDMA MR registration.

Also, as discussed with @jblomer, this is a separate problem that will be addressed in a separate PR if need be.

**Pull request closed without merging**